pkgs/development/libraries/matrix-sdk-crypto-nodejs/default.nix

```diff
@@ -0,0 +1,60 @@
+{ lib, stdenv, fetchFromGitHub, rustPlatform, napi-rs-cli, nodejs, libiconv }:
+
+stdenv.mkDerivation rec {
+  pname = "matrix-sdk-crypto-nodejs";
+  version = "0.1.0-beta.1";
+
+  src = fetchFromGitHub {
+    owner = "matrix-org";
+    repo = "matrix-rust-sdk";
+    rev = "${pname}-v${version}";
+    hash = "sha256-t9iaIfiw8m86Z0AvXgDaNVuy8odGSOf3gU/UtR6bevk=";
+  };
+
+  patches = [ ./cargo-lock.patch ];
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit src patches;
+    name = "${pname}-${version}";
+    hash = "sha256-5Z7co3kMiNMWFiRNBnJnT4iKlhGNQOiBFlrPcOySze0=";
+  };
+
+  postPatch = ''
+    cd bindings/${pname}
+  '';
+
+  nativeBuildInputs = [
+    rustPlatform.cargoSetupHook
+    rustPlatform.rust.cargo
+    rustPlatform.rust.rustc
+    napi-rs-cli
+    nodejs
+  ];
+
+  buildInputs = lib.optionals stdenv.isDarwin [ libiconv ];
+
+  buildPhase = ''
+    runHook preBuild
+
+    npm run release-build
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out/lib"
+    cp matrix-sdk-crypto.*.node "$out/lib"
```

Comment:
There should be an intermediary directory containing the `.node` files.
Perhaps `$out/lib/matrix-sdk-crypto-nodejs/`.

Reply:
I disagree -- why do you think this?

Reply:
If I recall correcly, this node package installs some native `.node` files
alongside an `index.js` which loads them from the same location or from known
node module paths.

As you [remarked], it could be a good idea to mimick the node output paths.

[remarked]: https://github.com/NixOS/nixpkgs/pull/191129#discussion_r970654422

Though this package only seems to generate the native `.node` files which are
then copied and merged with the other files of the same library in consuming
packages, so that's indeed not an usual library from which files are
automatically discovered.

Reply:
Probably best to copy the bindings too, and with it, move to `$out/lib/node_modules` since that now makes sense to use?

> so that's indeed not an usual library from which files are
automatically discovered

Was this agreeing with my point, or disagreeing with it?

Reply:
I think it would be better to produce and put the whole library under a
directory that the node environment can recorgine.

However, I'm not sure of how easy it is to override a dependency in an npm or
yarn package in nix to substitute it with this package (to avoid needing to
merge files manually). Maybe the complication isn't worth it. I'll be fine with
the current way.

Reply:
I can certainly bundle the generated JavaScript file(s) too, for the package to be complete, though we'd have to merge the files manually.